Handel Extensions Episode I: The Phantom ServiceRegistry

ThatJoeMoore opened the pull request:
Supports #302 , but doesn't fix it.

This PR doesn't actually add any features to Handel - yet.  However, it adds the concept of a 'Service Registry', and forces everything to use that instead of `util.getServiceDeployers()`.

The Service Registry knows how to load extensions from a list of extension definitions, but doesn't know how to download/install extensions or read a list of them from the handelfile.  That'll come in a later PR.  Like I said, this commit changes nothing about handel files or the public API, just the internal APIs that we use.

I'm going to go through the commit and make comments on some key changes.

ThatJoeMoore commented on extension-api/extension-api.ts:
```diff
@@ -77,13 +77,21 @@ export interface AccountConfig {
+export interface ServiceRegistry {
+    findDeployerFor(prefix: string, name: string): ServiceDeployer;
+    hasService(prefix: string, name: string): boolean;
+    validPrefixes(): Set<string>;
```
Hmm.  I'm gonna actually push a commit that makes these names better.  Gimme a few minutes.

ThatJoeMoore commented on extension-api/package.json:
```diff
@@ -10,11 +10,12 @@
   "license": "Apache-2.0",
-  "dependencies": {
+  "peerDependencies": {
     "aws-sdk": "^2.188.0"
```
By moving aws-sdk to a peer dependency, we can avoid having a million versions of aws-sdk running around inside of nested node_modules, which'll speed up our installs and startups. By like, a lot, because aws-sdk is HEAVY.

ThatJoeMoore commented on handel/gulpfile.js:
```diff
@@ -5,7 +5,6 @@ const fs = require('fs-extra');
 const path = require('path');
 const runSeq = require('run-sequence');
 const mergeStream = require('merge-stream');
-const mocha = require('gulp-spawn-mocha');
```
This wasn't actually getting used and was giving me console warnings, so I killed it 💣

ThatJoeMoore commented on extension-api/package.json:
```diff
@@ -10,11 +10,12 @@
   },
   "devDependencies": {
     "@types/node": "^9.4.0",
+    "aws-sdk": "^2.188.0",
```
...and to make Typescript happy, we have to install it as a dev dependency too 🤷‍♂️

ThatJoeMoore commented on handel/package.json:
```diff
@@ -18,7 +18,8 @@
     "archiver": "^1.3.0",
     "aws-sdk": "^2.169.0",
-    "fs-extra": "^4.0.2",
+    "common-tags": "^1.7.2",
```
This package gives us some nice utilities around tagged template literals, which can help us have nice pretty error messages in the extensions API (since I think that may get confusing to people if we don't have friendly error messages).

dsw88 commented on handel/src/datatypes/index.ts:
```diff
@@ -34,28 +34,15 @@ export interface AccountConfig extends api.AccountConfig {
-            this.params = params;
-            this.accountConfig = accountConfig;
-            this.tags = tags;
+    constructor(public appName: string,
```
Didn't even know you could do this. :)

ThatJoeMoore replied:
Right!?!

ThatJoeMoore commented on handel/test/service-registry/fake-service-registry.ts:
```diff
@@ -0,0 +1,44 @@
+    [key: string]: Partial<ServiceDeployer>;
+}
+
+export default class FakeServiceRegistry implements ServiceRegistry {
```
Sooo much easier to just have a fake object than to use Sinon for this.

dsw88 commented on handel/tsconfig.json:
```diff
@@ -1,7 +1,8 @@
   "compilerOptions": {
-    "allowJs": true
+    "allowJs": true,
+    "rootDir": "./"
```
I looked at the tsconfig docs for this, but I'm still a little unclear on what it does. Can you tell me how this changes the compilation?

ThatJoeMoore replied:
Among other things, it disallows relative paths that leave that directory, which is good because my IDE kept wanting to import the extension api from relative paths instead of from the node package.

ThatJoeMoore commented:
Per a face-to-face conversation with @dsw88 , I've removed ServiceRegistry from ServiceContext and replaced it with a ServiceInfo object.  That'll allow services to see information about their dependencies without having the ability to do unspeakable things with them.

dsw88 commented:
This is looking great so far!
